packages/contracts/src/contracts/current/protocol/Exchange/MixinExchangeCore.sol

```diff
@@ -77,134 +77,332 @@ contract MixinExchangeCore is
+            orderStatus = uint8(Status.INVALID);
+            return (orderStatus, orderHash, filledAmount);
+        }
 
-        // Check if order has been cancelled by salt value
-        if (order.salt < makerEpoch[order.makerAddress]) {
-            emit ExchangeError(uint8(Errors.ORDER_CANCELLED), orderHash);
-            return fillResults;
+        // If order.makerAssetAmount is zero, we also reject the order.
+        // While the Exchange contract handles them correctly, they create
+        // edge cases in the supporting infrastructure because they have
+        // an 'infinite' price when computed by a simple division.
+        if (order.makerAssetAmount == 0) {
+            orderStatus = uint8(Status.INVALID);
+            return (orderStatus, orderHash, filledAmount);
         }
 
-        // Check if order has been cancelled by orderHash
+        // Validate order expiration
+        if (block.timestamp >= order.expirationTimeSeconds) {
+            orderStatus = uint8(Status.ORDER_EXPIRED);
+            return (orderStatus, orderHash, filledAmount);
+        }
+
+        // Validate order availability
+        if (filledAmount >= order.takerAssetAmount) {
+            orderStatus = uint8(Status.ORDER_FULLY_FILLED);
+            return (orderStatus, orderHash, filledAmount);
+        }
+
+        // Check if order has been cancelled
         if (cancelled[orderHash]) {
-            emit ExchangeError(uint8(Errors.ORDER_CANCELLED), orderHash);
-            return fillResults;
+            orderStatus = uint8(Status.ORDER_CANCELLED);
+            return (orderStatus, orderHash, filledAmount);
+        }
+
+        // Validate order is not cancelled
+        if (makerEpoch[order.makerAddress] > order.salt) {
+            orderStatus = uint8(Status.ORDER_CANCELLED);
+            return (orderStatus, orderHash, filledAmount);
+        }
+
+        // Order is Fillable
+        orderStatus = uint8(Status.ORDER_FILLABLE);
+        return (orderStatus, orderHash, filledAmount);
+    }
+
+    /// @dev Validates context for fillOrder. Succeeds or throws.
+    /// @param order to be filled.
+    /// @param orderStatus Status of order to be filled.
+    /// @param orderHash Hash of order to be filled.
+    /// @param filledAmount Amount of order already filled.
+    /// @param signature Proof that the orders was created by its maker.
+    /// @param takerAddress Address of order taker.
+    /// @param takerAssetFillAmount Desired amount of order to fill by taker.
+    function validateFillOrderContextOrRevert(
+        Order memory order,
+        uint8 orderStatus,
+        bytes32 orderHash,
+        uint256 filledAmount,
+        bytes memory signature,
+        address takerAddress,
+        uint256 takerAssetFillAmount)
+        internal
+    {
+        // Ensure order status is not invalid
+        if (orderStatus == uint8(Status.INVALID)) {
+            emit ExchangeStatus(uint8(orderStatus), orderHash);
```

Comment:
`revert` reverts all state, including events, so this patten won't work. Instead, let's just use a `require` with a reason.

Reply:
👍  I'll replace with `require(..., [message])` after rebasing against `v0.4.23`